tensorflow/lite/micro/micro_profiler.h

```diff
@@ -33,67 +46,185 @@ class MicroProfiler : public MicroProfilerInterface {
+                      tag_groups_[i].ticks, us / 1000, us % 1000);
+        }
+        us = TicksToUs(total_ticks);
+        MicroPrintf("\nTotal time: %" PRIu64 ".%03" PRIu64 " ms (%lld ticks)",
+                    us / 1000, us % 1000, total_ticks);
+        break;
+      }
+      case MicroProfilerLogFormat::Csv: {
+        MicroPrintf("\"Tag\",\"Total ms\",\"Total ticks\"");
+        for (int i = 0; i < num_tag_groups_; ++i) {
+          uint64_t us = TicksToUs(tag_groups_[i].ticks);
+          MicroPrintf("%s, %" PRIu64 ".%" PRIu64 ", %u", tag_groups_[i].tag,
+                      us / 1000, us % 1000, tag_groups_[i].ticks);
+        }
+        break;
+      }
+    }
+#endif  // !defined(TF_LITE_STRIP_ERROR_STRINGS)
+  }
 
  private:
-  // Maximum number of events that this class can keep track of. The
-  // MicroProfiler will abort if AddEvent is called more than kMaxEvents number
-  // of times. Increase this number if you need more events.
-  static constexpr int kMaxEvents = 4096;
-
-  const char* tags_[kMaxEvents];
-  uint32_t start_ticks_[kMaxEvents];
-  uint32_t end_ticks_[kMaxEvents];
+  const char* tags_[MAX_EVENTS];
+  uint32_t start_ticks_[MAX_EVENTS];
+  uint32_t end_ticks_[MAX_EVENTS];
   int num_events_ = 0;
+  int num_tag_groups_ = 0;
 
-  struct TicksPerTag {
+  struct TagGroup {
     const char* tag;
     uint32_t ticks;
+    uint32_t tag_count;
   };
+
   // In practice, the number of tags will be much lower than the number of
   // events. But it is theoretically possible that each event to be unique and
-  // hence we allow total_ticks_per_tag to have kMaxEvents entries.
-  TicksPerTag total_ticks_per_tag[kMaxEvents] = {};
+  // hence we allow total_ticks_per_tag to have MAX_EVENTS entries.
+  TagGroup tag_groups_[MAX_EVENTS] = {};
+
+  // Helper function to find the index of a tag in the cumulative array
+  TagGroup& GetTagGroup(const char* tag) {
+    for (int i = 0; i < num_tag_groups_; ++i) {
+      if (strcmp(tag_groups_[i].tag, tag) == 0) {
+        return tag_groups_[i];
+      }
+    }
 
-  int FindExistingOrNextPosition(const char* tag_name);
+    // Tag not found, so we create a new entry
+    // There should always be space since the array of tag groups
+    // is just as big as the array of events
+    tag_groups_[num_tag_groups_].tag = tag;
+    tag_groups_[num_tag_groups_].ticks = 0;
+    tag_groups_[num_tag_groups_].tag_count = 0;
+    return tag_groups_[num_tag_groups_++];
+  }
+
+  // Helper function to sort the tag groups by ticks in descending order
+  // Simple bubble sort implementation
+  void SortTagGroups() {
+    for (int i = 0; i < num_tag_groups_ - 1; ++i) {
+      for (int j = i + 1; j < num_tag_groups_; ++j) {
+        if (tag_groups_[j].ticks > tag_groups_[i].ticks) {
+          TagGroup temp_tag_group = tag_groups_[i];
+          tag_groups_[i] = tag_groups_[j];
+          tag_groups_[j] = temp_tag_group;
+        }
+      }
+    }
+  }
```

Comment:
Please use standard C++ library algorithms (as long as they don't require memory allocation).  Here `std::sort` should be used.

Reply:
Done